# Predicting Absenteeism from Work

## Data Preprocessing

In [1]:
import pandas as pd

In [2]:
raw=pd.read_csv("Absenteeism_data.csv")
raw.head(5)

ID  Reason for Absence        Date  Transportation Expense  \
0  11                  26  07/07/2015                     289   
1  36                   0  14/07/2015                     118   
2   3                  23  15/07/2015                     179   
3   7                   7  16/07/2015                     279   
4  11                  23  23/07/2015                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  
0         2     1                          4  
1         1     0                          0  
2         0     0                          2  
3         2     0                          4  
4         2     1                          2

In [3]:
df=raw.copy()
df

ID  Reason for Absence        Date  Transportation Expense  \
0    11                  26  07/07/2015                     289   
1    36                   0  14/07/2015                     118   
2     3                  23  15/07/2015                     179   
3     7                   7  16/07/2015                     279   
4    11                  23  23/07/2015                     289   
..   ..                 ...         ...                     ...   
695  17                  10  23/05/2018                     179   
696  28                   6  23/05/2018                     225   
697  18                  10  24/05/2018                     330   
698  25                  23  24/05/2018                     235   
699  15                  28  31/05/2018                     291   

     Distance to Work  Age  Daily Work Load Average  Body Mass Index  \
0                  36   33                  239.554               30   
1                  13   50                  239.554               31   
2                  51   38                  239.554               31   
3                   5   39                  239.554               24   
4                  36   33                  239.554               30   
..                ...  ...                      ...              ...   
695                22   40                  237.656               22   
696                26   28                  237.656               24   
697                16   28                  237.656               25   
698                16   32                  237.656               25   
699                31   40                  237.656               25   

     Education  Children  Pets  Absenteeism Time in Hours  
0            1         2     1                          4  
1            1         1     0                          0  
2            1         0     0                          2  
3            1         2     0                          4  
4            1         2     1                          2  
..         ...       ...   ...                        ...  
695          2         2     0                          8  
696          1         1     2                          3  
697          2         0     0                          8  
698          3         0     0                          2  
699          1         1     1                          2  

[700 rows x 12 columns]

In [4]:
#Setting Display Options

#pd.options.display.max_columns=None
pd.options.display.max_rows=10

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ID                         700 non-null    int64  
 1   Reason for Absence         700 non-null    int64  
 2   Date                       700 non-null    object 
 3   Transportation Expense     700 non-null    int64  
 4   Distance to Work           700 non-null    int64  
 5   Age                        700 non-null    int64  
 6   Daily Work Load Average    700 non-null    float64
 7   Body Mass Index            700 non-null    int64  
 8   Education                  700 non-null    int64  
 9   Children                   700 non-null    int64  
 10  Pets                       700 non-null    int64  
 11  Absenteeism Time in Hours  700 non-null    int64  
dtypes: float64(1), int64(10), object(1)
memory usage: 65.8+ KB


700 rows in each column mean there are no missing values.

Dependent Variable: Absenteeism Time in Hours

In [6]:
# Dropping the ID column

df=df.drop(['ID'], axis=1)
df.head(5)

Reason for Absence        Date  Transportation Expense  Distance to Work  \
0                  26  07/07/2015                     289                36   
1                   0  14/07/2015                     118                13   
2                  23  15/07/2015                     179                51   
3                   7  16/07/2015                     279                 5   
4                  23  23/07/2015                     289                36   

   Age  Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0   33                  239.554               30          1         2     1   
1   50                  239.554               31          1         1     0   
2   38                  239.554               31          1         0     0   
3   39                  239.554               24          1         2     0   
4   33                  239.554               30          1         2     1   

   Absenteeism Time in Hours  
0                          4  
1                          0  
2                          2  
3                          4  
4                          2

### Examining the Reason for Absence

In [7]:
df['Reason for Absence']

0      26
1       0
2      23
3       7
4      23
       ..
695    10
696     6
697    10
698    23
699    28
Name: Reason for Absence, Length: 700, dtype: int64

In [8]:
df['Reason for Absence'].min()

0

In [9]:
df['Reason for Absence'].min()

0

In [10]:
pd.unique(df['Reason for Absence'])

# df['Reason for Absence'].unique()

array([26,  0, 23,  7, 22, 19,  1, 11, 14, 21, 10, 13, 28, 18, 25, 24,  6,
       27, 17,  8, 12,  5,  9, 15,  4,  3,  2, 16])

In [11]:
len(df['Reason for Absence'].unique())

28

In [12]:
#Sorting the list

sorted(df['Reason for Absence'].unique())

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28]

##### We have found that reason#20 is missing

#### Dummy Variables:

1 if a certain categorical effect is present, 0 if it is absent

In [13]:
#To ensure that the employees were absent for only ONE reason.

reason_columns=pd.get_dummies(df['Reason for Absence'])
reason_columns

0      1      2      3      4      5      6      7      8      9   \
0    False  False  False  False  False  False  False  False  False  False   
1     True  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False  False   True  False  False   
4    False  False  False  False  False  False  False  False  False  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
695  False  False  False  False  False  False  False  False  False  False   
696  False  False  False  False  False  False   True  False  False  False   
697  False  False  False  False  False  False  False  False  False  False   
698  False  False  False  False  False  False  False  False  False  False   
699  False  False  False  False  False  False  False  False  False  False   

     ...     18     19     21     22     23     24     25     26     27     28  
0    ...  False  False  False  False  False  False  False   True  False  False  
1    ...  False  False  False  False  False  False  False  False  False  False  
2    ...  False  False  False  False   True  False  False  False  False  False  
3    ...  False  False  False  False  False  False  False  False  False  False  
4    ...  False  False  False  False   True  False  False  False  False  False  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
695  ...  False  False  False  False  False  False  False  False  False  False  
696  ...  False  False  False  False  False  False  False  False  False  False  
697  ...  False  False  False  False  False  False  False  False  False  False  
698  ...  False  False  False  False   True  False  False  False  False  False  
699  ...  False  False  False  False  False  False  False  False  False   True  

[700 rows x 28 columns]

In [14]:
#Checking Rows by ensuring the same (no repition of reasons)
#0 means missing value
#1 means single value
#2 or higher means repitition

reason_columns['check']=reason_columns.sum(axis=1)
reason_columns


0      1      2      3      4      5      6      7      8      9  \
0    False  False  False  False  False  False  False  False  False  False   
1     True  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False  False   True  False  False   
4    False  False  False  False  False  False  False  False  False  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
695  False  False  False  False  False  False  False  False  False  False   
696  False  False  False  False  False  False   True  False  False  False   
697  False  False  False  False  False  False  False  False  False  False   
698  False  False  False  False  False  False  False  False  False  False   
699  False  False  False  False  False  False  False  False  False  False   

     ...     19     21     22     23     24     25     26     27     28  check  
0    ...  False  False  False  False  False  False   True  False  False      1  
1    ...  False  False  False  False  False  False  False  False  False      1  
2    ...  False  False  False   True  False  False  False  False  False      1  
3    ...  False  False  False  False  False  False  False  False  False      1  
4    ...  False  False  False   True  False  False  False  False  False      1  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
695  ...  False  False  False  False  False  False  False  False  False      1  
696  ...  False  False  False  False  False  False  False  False  False      1  
697  ...  False  False  False  False  False  False  False  False  False      1  
698  ...  False  False  False   True  False  False  False  False  False      1  
699  ...  False  False  False  False  False  False  False  False   True      1  

[700 rows x 29 columns]

In [15]:
#To check the sum of values in the check column. It must be 700.
reason_columns['check'].sum(axis=0)

700

In [16]:
reason_columns['check'].unique()

array([1])

In [17]:
reason_columns=reason_columns.drop(['check'],axis=1)
reason_columns

0      1      2      3      4      5      6      7      8      9   \
0    False  False  False  False  False  False  False  False  False  False   
1     True  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False  False   True  False  False   
4    False  False  False  False  False  False  False  False  False  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
695  False  False  False  False  False  False  False  False  False  False   
696  False  False  False  False  False  False   True  False  False  False   
697  False  False  False  False  False  False  False  False  False  False   
698  False  False  False  False  False  False  False  False  False  False   
699  False  False  False  False  False  False  False  False  False  False   

     ...     18     19     21     22     23     24     25     26     27     28  
0    ...  False  False  False  False  False  False  False   True  False  False  
1    ...  False  False  False  False  False  False  False  False  False  False  
2    ...  False  False  False  False   True  False  False  False  False  False  
3    ...  False  False  False  False  False  False  False  False  False  False  
4    ...  False  False  False  False   True  False  False  False  False  False  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
695  ...  False  False  False  False  False  False  False  False  False  False  
696  ...  False  False  False  False  False  False  False  False  False  False  
697  ...  False  False  False  False  False  False  False  False  False  False  
698  ...  False  False  False  False   True  False  False  False  False  False  
699  ...  False  False  False  False  False  False  False  False  False   True  

[700 rows x 28 columns]

#### Avoiding Multicolinearity

In [18]:
#Dropping Column 0

reason_columns=pd.get_dummies(df['Reason for Absence'],drop_first=True)
reason_columns

1      2      3      4      5      6      7      8      9      10  \
0    False  False  False  False  False  False  False  False  False  False   
1    False  False  False  False  False  False  False  False  False  False   
2    False  False  False  False  False  False  False  False  False  False   
3    False  False  False  False  False  False   True  False  False  False   
4    False  False  False  False  False  False  False  False  False  False   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
695  False  False  False  False  False  False  False  False  False   True   
696  False  False  False  False  False   True  False  False  False  False   
697  False  False  False  False  False  False  False  False  False   True   
698  False  False  False  False  False  False  False  False  False  False   
699  False  False  False  False  False  False  False  False  False  False   

     ...     18     19     21     22     23     24     25     26     27     28  
0    ...  False  False  False  False  False  False  False   True  False  False  
1    ...  False  False  False  False  False  False  False  False  False  False  
2    ...  False  False  False  False   True  False  False  False  False  False  
3    ...  False  False  False  False  False  False  False  False  False  False  
4    ...  False  False  False  False   True  False  False  False  False  False  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
695  ...  False  False  False  False  False  False  False  False  False  False  
696  ...  False  False  False  False  False  False  False  False  False  False  
697  ...  False  False  False  False  False  False  False  False  False  False  
698  ...  False  False  False  False   True  False  False  False  False  False  
699  ...  False  False  False  False  False  False  False  False  False   True  

[700 rows x 27 columns]

#### Grouping the reasons for absence

In [19]:
#Dataframe columns

df.columns.values

array(['Reason for Absence', 'Date', 'Transportation Expense',
       'Distance to Work', 'Age', 'Daily Work Load Average',
       'Body Mass Index', 'Education', 'Children', 'Pets',
       'Absenteeism Time in Hours'], dtype=object)

In [20]:
#Reasons Dummy Columns
reason_columns.columns.values

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 21, 22, 23, 24, 25, 26, 27, 28])

In [21]:
df=df.drop(['Reason for Absence'],axis=1)
df

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  
0                            4  
1                            0  
2                            2  
3                            4  
4                            2  
..                         ...  
695                          8  
696                          3  
697                          8  
698                          2  
699                          2  

[700 rows x 10 columns]

In [22]:
#Grouping the Reasons into similar classes using .loc[] method

reas1=reason_columns.loc[:, 1:14].max(axis=1)
reas2=reason_columns.loc[:, 15:17].max(axis=1)
reas3=reason_columns.loc[:, 18:21].max(axis=1)
reas4=reason_columns.loc[:, 22:].max(axis=1)

#### Concatenate the data

In [23]:
df=pd.concat([df,reas1,reas2,reas3,reas4],axis=1)
df

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours      0      1      2      3  
0                            4  False  False  False   True  
1                            0  False  False  False  False  
2                            2  False  False  False   True  
3                            4   True  False  False  False  
4                            2  False  False  False   True  
..                         ...    ...    ...    ...    ...  
695                          8   True  False  False  False  
696                          3   True  False  False  False  
697                          8   True  False  False  False  
698                          2  False  False  False   True  
699                          2  False  False  False   True  

[700 rows x 14 columns]

#### Assigning Column Names

In [24]:
df.columns.values

array(['Date', 'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Absenteeism Time in Hours', 0, 1, 2, 3],
      dtype=object)

In [25]:
column_names=['Date', 'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Absenteeism Time in Hours', 'Reason 1', 'Reason 2', 'Reason 3', 'Reason 4']

df.columns=column_names
df

Date  Transportation Expense  Distance to Work  Age  \
0    07/07/2015                     289                36   33   
1    14/07/2015                     118                13   50   
2    15/07/2015                     179                51   38   
3    16/07/2015                     279                 5   39   
4    23/07/2015                     289                36   33   
..          ...                     ...               ...  ...   
695  23/05/2018                     179                22   40   
696  23/05/2018                     225                26   28   
697  24/05/2018                     330                16   28   
698  24/05/2018                     235                16   32   
699  31/05/2018                     291                31   40   

     Daily Work Load Average  Body Mass Index  Education  Children  Pets  \
0                    239.554               30          1         2     1   
1                    239.554               31          1         1     0   
2                    239.554               31          1         0     0   
3                    239.554               24          1         2     0   
4                    239.554               30          1         2     1   
..                       ...              ...        ...       ...   ...   
695                  237.656               22          2         2     0   
696                  237.656               24          1         1     2   
697                  237.656               25          2         0     0   
698                  237.656               25          3         0     0   
699                  237.656               25          1         1     1   

     Absenteeism Time in Hours  Reason 1  Reason 2  Reason 3  Reason 4  
0                            4     False     False     False      True  
1                            0     False     False     False     False  
2                            2     False     False     False      True  
3                            4      True     False     False     False  
4                            2     False     False     False      True  
..                         ...       ...       ...       ...       ...  
695                          8      True     False     False     False  
696                          3      True     False     False     False  
697                          8      True     False     False     False  
698                          2     False     False     False      True  
699                          2     False     False     False      True  

[700 rows x 14 columns]

#### Reordering the columns

In [26]:
column_names_reordered = ['Reason 1', 'Reason 2', 'Reason 3', 'Reason 4','Date', 'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education',
       'Children', 'Pets', 'Absenteeism Time in Hours']

df=df[column_names_reordered]

df.head(5)

Reason 1  Reason 2  Reason 3  Reason 4        Date  Transportation Expense  \
0     False     False     False      True  07/07/2015                     289   
1     False     False     False     False  14/07/2015                     118   
2     False     False     False      True  15/07/2015                     179   
3      True     False     False     False  16/07/2015                     279   
4     False     False     False      True  23/07/2015                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  
0         2     1                          4  
1         1     0                          0  
2         0     0                          2  
3         2     0                          4  
4         2     1                          2

#### Creating a Checkpoint

In [27]:
df_reason_mod=df.copy()
df_reason_mod

Reason 1  Reason 2  Reason 3  Reason 4        Date  \
0       False     False     False      True  07/07/2015   
1       False     False     False     False  14/07/2015   
2       False     False     False      True  15/07/2015   
3        True     False     False     False  16/07/2015   
4       False     False     False      True  23/07/2015   
..        ...       ...       ...       ...         ...   
695      True     False     False     False  23/05/2018   
696      True     False     False     False  23/05/2018   
697      True     False     False     False  24/05/2018   
698     False     False     False      True  24/05/2018   
699     False     False     False      True  31/05/2018   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          1         2     1                          4  
1                 31          1         1     0                          0  
2                 31          1         0     0                          2  
3                 24          1         2     0                          4  
4                 30          1         2     1                          2  
..               ...        ...       ...   ...                        ...  
695               22          2         2     0                          8  
696               24          1         1     2                          3  
697               25          2         0     0                          8  
698               25          3         0     0                          2  
699               25          1         1     1                          2  

[700 rows x 14 columns]

#### Working the Dates

In [28]:
df_reason_mod['Date']

0      07/07/2015
1      14/07/2015
2      15/07/2015
3      16/07/2015
4      23/07/2015
          ...    
695    23/05/2018
696    23/05/2018
697    24/05/2018
698    24/05/2018
699    31/05/2018
Name: Date, Length: 700, dtype: object

In [29]:
type(df_reason_mod['Date'])

pandas.core.series.Series

In [30]:
type(df_reason_mod['Date'][0])

str

In [31]:
df_reason_mod['Date'] = pd.to_datetime(df_reason_mod['Date'], format='%d/%m/%Y')

In [32]:
pd.options.display.max_rows=None

In [33]:
df_reason_mod['Date']

0     2015-07-07
1     2015-07-14
2     2015-07-15
3     2015-07-16
4     2015-07-23
5     2015-07-10
6     2015-07-17
7     2015-07-24
8     2015-07-06
9     2015-07-13
10    2015-07-20
11    2015-07-14
12    2015-07-15
13    2015-07-15
14    2015-07-15
15    2015-07-17
16    2015-07-17
17    2015-07-27
18    2015-07-30
19    2015-08-05
20    2015-08-12
21    2015-08-03
22    2015-08-10
23    2015-08-14
24    2015-08-17
25    2015-08-24
26    2015-08-04
27    2015-08-12
28    2015-08-19
29    2015-08-28
30    2015-08-17
31    2015-08-27
32    2015-08-27
33    2015-08-17
34    2015-08-17
35    2015-08-17
36    2015-08-04
37    2015-08-20
38    2015-08-21
39    2015-08-28
40    2015-09-01
41    2015-09-07
42    2015-09-01
43    2015-09-08
44    2015-09-09
45    2015-09-13
46    2015-09-14
47    2015-09-24
48    2015-09-04
49    2015-09-14
50    2015-09-21
51    2015-09-28
52    2015-09-08
53    2015-09-15
54    2015-09-22
55    2015-09-29
56    2015-09-16
57    2015-09-23
58    2015-09-

In [34]:
pd.options.display.max_rows=5

In [35]:
df_reason_mod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   Reason 1                   700 non-null    bool          
 1   Reason 2                   700 non-null    bool          
 2   Reason 3                   700 non-null    bool          
 3   Reason 4                   700 non-null    bool          
 4   Date                       700 non-null    datetime64[ns]
 5   Transportation Expense     700 non-null    int64         
 6   Distance to Work           700 non-null    int64         
 7   Age                        700 non-null    int64         
 8   Daily Work Load Average    700 non-null    float64       
 9   Body Mass Index            700 non-null    int64         
 10  Education                  700 non-null    int64         
 11  Children                   700 non-null    int64         
 12  Pets    

#### Extracting the Month from Dates

In [36]:
df_reason_mod['Date'][0]

Timestamp('2015-07-07 00:00:00')

In [37]:
df_reason_mod['Date'][0].month

7

In [38]:
list_months=[]
list_months

[]

In [39]:
df_reason_mod.shape

(700, 14)

In [40]:
for i in range(df_reason_mod.shape[0]):
    list_months.append(df_reason_mod['Date'][i].month)

In [41]:
list_months

[7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 7,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 9,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 10,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 11,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 12,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 7,
 

In [42]:
df_reason_mod['Month Value']=list_months
df_reason_mod

Reason 1  Reason 2  Reason 3  Reason 4       Date  \
0       False     False     False      True 2015-07-07   
1       False     False     False     False 2015-07-14   
..        ...       ...       ...       ...        ...   
698     False     False     False      True 2018-05-24   
699     False     False     False      True 2018-05-31   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
..                      ...               ...  ...                      ...   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0                 30          1         2     1                          4   
1                 31          1         1     0                          0   
..               ...        ...       ...   ...                        ...   
698               25          3         0     0                          2   
699               25          1         1     1                          2   

     Month Value  
0              7  
1              7  
..           ...  
698            5  
699            5  

[700 rows x 15 columns]

#### Extracting Day of the Week

In [43]:
df_reason_mod['Date'][699].weekday()

3

In [44]:
#Creating a function
def date_to_weekday(date_value):
    return date_value.weekday()

In [45]:
df_reason_mod['Day of the Week']=df_reason_mod['Date'].apply(date_to_weekday)

In [46]:
df_reason_mod.head(5)

Reason 1  Reason 2  Reason 3  Reason 4       Date  Transportation Expense  \
0     False     False     False      True 2015-07-07                     289   
1     False     False     False     False 2015-07-14                     118   
2     False     False     False      True 2015-07-15                     179   
3      True     False     False     False 2015-07-16                     279   
4     False     False     False      True 2015-07-23                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  Month Value  Day of the Week  
0         2     1                          4            7                1  
1         1     0                          0            7                1  
2         0     0                          2            7                2  
3         2     0                          4            7                3  
4         2     1                          2            7                3

After we created 4 reason type columns, we filled them with dummy values only. We dropped the ‘Reason for Absence’ column as the information it conveyed was contained in the newly created columns as well.

In [47]:
df_mondate=df_reason_mod.copy()
df_mondate

Reason 1  Reason 2  Reason 3  Reason 4       Date  \
0       False     False     False      True 2015-07-07   
1       False     False     False     False 2015-07-14   
..        ...       ...       ...       ...        ...   
698     False     False     False      True 2018-05-24   
699     False     False     False      True 2018-05-31   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
..                      ...               ...  ...                      ...   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0                 30          1         2     1                          4   
1                 31          1         1     0                          0   
..               ...        ...       ...   ...                        ...   
698               25          3         0     0                          2   
699               25          1         1     1                          2   

     Month Value  Day of the Week  
0              7                1  
1              7                1  
..           ...              ...  
698            5                3  
699            5                3  

[700 rows x 16 columns]

In [48]:
#Dropping the Date column

df_mondate=df_mondate.drop(['Date'],axis=1)
df_mondate

Reason 1  Reason 2  Reason 3  Reason 4  Transportation Expense  \
0       False     False     False      True                     289   
1       False     False     False     False                     118   
..        ...       ...       ...       ...                     ...   
698     False     False     False      True                     235   
699     False     False     False      True                     291   

     Distance to Work  Age  Daily Work Load Average  Body Mass Index  \
0                  36   33                  239.554               30   
1                  13   50                  239.554               31   
..                ...  ...                      ...              ...   
698                16   32                  237.656               25   
699                31   40                  237.656               25   

     Education  Children  Pets  Absenteeism Time in Hours  Month Value  \
0            1         2     1                          4            7   
1            1         1     0                          0            7   
..         ...       ...   ...                        ...          ...   
698          3         0     0                          2            5   
699          1         1     1                          2            5   

     Day of the Week  
0                  1  
1                  1  
..               ...  
698                3  
699                3  

[700 rows x 15 columns]

In [49]:
df_mondate.columns

Index(['Reason 1', 'Reason 2', 'Reason 3', 'Reason 4',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education', 'Children',
       'Pets', 'Absenteeism Time in Hours', 'Month Value', 'Day of the Week'],
      dtype='object')

In [50]:
column_mod=['Reason 1', 'Reason 2', 'Reason 3', 'Reason 4', 'Month Value', 'Day of the Week',
       'Transportation Expense', 'Distance to Work', 'Age',
       'Daily Work Load Average', 'Body Mass Index', 'Education', 'Children',
       'Pets', 'Absenteeism Time in Hours']

df_mondate=df_mondate[column_mod]

df_mondate.head(5)

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0     False     False     False      True            7                1   
1     False     False     False     False            7                1   
2     False     False     False      True            7                2   
3      True     False     False     False            7                3   
4     False     False     False      True            7                3   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          1         2     1                          4  
1               31          1         1     0                          0  
2               31          1         0     0                          2  
3               24          1         2     0                          4  
4               30          1         2     1                          2

#### Creating Dummy Variables of Education

In [51]:
df_mondate['Education'].unique()

array([1, 3, 2, 4])

In [52]:
df_mondate['Education'].value_counts()

Education
1    583
3     73
2     40
4      4
Name: count, dtype: int64

As seen above, 583 employees have high school education and only a little over 100 have higher education, so we have categorised them into two variables

In [53]:
df_mondate['Education']=df_mondate['Education'].map({1:0,2:1,3:1,4:1})

In [54]:
df_mondate['Education'].unique()

array([0, 1])

In [55]:
df_mondate['Education'].value_counts()

Education
0    583
1    117
Name: count, dtype: int64

In [56]:
df_preprocessed=df_mondate.copy()
df_preprocessed.to_csv('df_preprocessed.csv', index=False)
df_preprocessed

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0       False     False     False      True            7                1   
1       False     False     False     False            7                1   
..        ...       ...       ...       ...          ...              ...   
698     False     False     False      True            5                3   
699     False     False     False      True            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
..                      ...               ...  ...                      ...   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          0         2     1                          4  
1                 31          0         1     0                          0  
..               ...        ...       ...   ...                        ...  
698               25          1         0     0                          2  
699               25          0         1     1                          2  

[700 rows x 15 columns]